src/testdir/test_ins_complete.vim

```diff
@@ -2764,24 +2767,54 @@ func Test_complete_fuzzy_match()
   call assert_equal('你好 他好', getline('.'))
 
   " issue #15526
-  set completeopt=fuzzy,menuone,menu,noselect
+  set completeopt=menuone,menu,noselect
   call setline(1, ['Text', 'ToText', ''])
   call cursor(3, 1)
   call feedkeys("STe\<C-X>\<C-N>x\<CR>\<Esc>0", 'tx!')
   call assert_equal('Tex', getline('.'))
 
-  " clean up
-  set omnifunc=
   bw!
   bw!
-  set complete& completeopt&
-  autocmd! AAAAA_Group
-  augroup! AAAAA_Group
-  delfunc OnPumChange
-  delfunc Omni_test
-  delfunc Comp
-  unlet g:item
-  unlet g:word
+  set completeopt& cfc& cpt&
+endfunc
+
+func Test_cfc_with_longest()
+  new
+  set completefuzzycollect=k,f,l
+  set completeopt=menu,menuone,longest,fuzzy
+
+  " keyword
+  exe "normal ggdGShello helio think h\<C-X>\<C-N>\<esc>"
+  call assert_equal("hello helio think hel", getline('.'))
+
+  " file
+  call writefile([''], 'hello', 'D')
+  call writefile([''], 'helio', 'D')
+  call writefile([''], 'think', 'D')
+  call feedkeys("Sh\<C-X>\<C-f>\<Esc>0", 'tx!')
+  call assert_equal('hel', getline('.'))
+
+  " word
+  call setline(1, ['what do you think', 'why i have that', ''])
+  call cursor(3,1)
+  call feedkeys("Sw\<C-X>\<C-l>\<C-N>\<Esc>0", 'tx!')
+  call assert_equal('wh', getline('.'))
+
+  exe "normal ggdG"
+  " auto complete when only one match
+  exe "normal Shello\<CR>h\<C-X>\<C-N>\<esc>"
+  call assert_equal('hello', getline('.'))
```

Comment:
@roccomao  I have added a case for auto complete when only have one result. and below is about search with new input. how do you feel ?

Reply:
To simplify the description, you can change line 2805 to any of the following. The three writings are equivalent and the test will not pass.

```vim
  exe "normal Shello\<CR>h\<C-X>\<C-N>\<Esc>cch\<C-X>\<C-N>\<Esc>"
  exe "normal Shello\<CR>h\<C-X>\<C-N>\<C-W>h\<C-X>\<C-N>\<Esc>"
  exe "normal Shello\<CR>h\<C-X>\<C-N>\<Esc>oh\<C-X>\<C-N>\<Esc>"
```

Reply:
The same issue when the keyword is from a file, described by the test case is as follows:

```vim
  exe "normal! ggdG"
  call writefile(['hello'], 'test_keyword.txt', 'D')
  set complete+=ktest_keyword.txt
  exe "normal! Sh\<C-N>\<Esc>"
  call assert_equal('hello', getline('.'))
```

Still the same as before, we need to press `<C-n>` twice, So following test will pass, which shouldn't be :

```vim
  exe "normal! ggdG"
  call writefile(['hello'], 'test_keyword.txt', 'D')
  set complete+=ktest_keyword.txt
  exe "normal! Sh\<C-N>\<C-N>\<Esc>"
  call assert_equal('hello', getline('.'))
```

Reply:
kdict not support yet..

Reply:
Oh, indeed, I just discovered this problem during use to test. I tried it again just now and it is indeed impossible to fuzzy collect the words in `k{dict}`. My mistake.

Reply:
not any mistake here :)  most important thing is how do you feel of current `longest(fuzzy)` + `fuzzycollect` behavior ?  if okay i will continue work on kdict and other compl type.

Reply:
I personally think it's pretty good.

Reply:
Regarding the performance loss you mentioned earlier, how much will be lost exactly? Is there any way to test this?

Reply:
good to know, I will continue works one other compl type of keyword then

Reply:
> Regarding the performance loss you mentioned earlier, how much will be lost exactly? Is there any way to test this?

This is an obvious performance issue. Because the first doubly linked list traversal does not find any matches, when the condition is met, compl_leader will be regenerated and the doubly linked list will be traversed again. O(n) x O(n). So as long as n is large enough, you should be able to feel some delay.

Reply:
I'm not particularly concerned about this, nor do I know much about the implementation details. I'm just curious. Thanks for your answer.

Reply:
now kdict should work and i have add some simple cases for kdict maybe they are not enough to covers most situations. if you got problem let me know 👍